syntax = "proto3";

import "google/protobuf/any.proto";

package protocol;

message Transaction {
    message Contract {
        enum ContractType {
            AccountCreateContract = 0;
            TransferContract = 1;
            TransferAssetContract = 2;
        }
        ContractType type = 1;
        google.protobuf.Any parameter = 2;
        bytes provider = 3;
    }

    message raw {
        bytes ref_block_bytes = 1;
        int64 ref_block_num = 3;
        bytes ref_block_hash = 4;
        int64 expiration = 8;
        //only support size = 1,  repeated list here for extension
        repeated Contract contract = 11;
        int64 timestamp = 14;
        int64 fee_limit = 18;
    }

    raw raw_data = 1;
}

message BlockHeader {
    message raw {
        int64 timestamp = 1;
        bytes tx_trie_root = 2;
        bytes parent_hash = 3;
        int64 number = 7;
        int64 witness_id = 8;
        bytes witness_address = 9;
        int32 version = 10;
    }
    raw raw_data = 1;
    bytes witness_signature = 2;
}

message TransferContract {
    bytes owner_address = 1;
    bytes to_address = 2;
    int64 amount = 3;
}

message TransferAssetContract {
    bytes asset_name = 1;
    bytes owner_address = 2;
    bytes to_address = 3;
    int64 amount = 4;
}
